### PR TITLE
New gateway: HSBC Direct Debit API for HK

### PR DESCRIPTION
**Pull request by bruno**
This Pull Request introduces a new payment gateway (HSBC HK) that *only* supports an API for direct debit authorisations. This is only for Hong Kong.

We've also introduced a `Dockerfile` and associated tooling to enable local development using docker.

### CODE REVIEW

**Review by jscheid — COMMENTED** [review]
Nice work Henare. I worry a little about the crypto performance, considering that it appears to be used per-request. Do you have an idea of its overhead?

**Review by bruno — COMMENTED** [review]
Great job on taking this @henare ... It does look good to me.

**Comment by henare** [comment]
> Nice work Henare. I worry a little about the crypto performance, considering that it appears to be used per-request. Do you have an idea of its overhead?

Good point @jscheid - we haven't tested crypto performance at all so far. I also note this implementation writes to disk as part of the crypto operations.

I just did some crude tests and on my machine and it seems to take between 60 and 90ms to encrypt a payload. Decrypting was sub-millisecond. Given that, is it something we should look into more?

**Comment by fbrusatti** [comment]
Outstanding work here guys 👏 . 

If performance of encryption/decryption is an issue, a very small tweak could be decrypt only once in `commit` method and avoid decrypt twice (because `message_from` where you decrypt is called twice)

⚠️ remember before push this to change the `live_url` (I did that mistake before)

File: lib/active_merchant/billing/gateways/hsbc.rb [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      self.display_name = 'HSBC'
+
+      STANDARD_ERROR_CODE_MAPPING = {}
+
+      def initialize(options={})
+        requires!(options, :client_id, :client_secret, :profile_id, :public_key, :private_key)
+        super
+      end
+
+      def authorize(money, payment, options={})
```
Comment (fbrusatti) [inline lib/active_merchant/billing/gateways/hsbc.rb]: is `payment` needed here? Is it going to be used in a future? I think we don't use it in anyplace now.

**Comment by jscheid** [comment]
I reckon 60-90ms overhead per payment isn't great, but acceptable. Maybe something that could be improved down the road. If it's multiples of 60-90ms per payment (not sure how many requests there are and how many of them need encryption) then maybe it's worth looking into it sooner than later.

File: lib/active_merchant/billing/gateways/hsbc.rb (hunk 1 above)
Comment (henare) [reply lib/active_merchant/billing/gateways/hsbc.rb]: Nice catch! Fixed in https://github.com/waysact/active_merchant/pull/157/commits/09e9d5a9f704fe403e23459e41aa592781aa9046.

**Comment by henare** [comment]
> If performance of encryption/decryption is an issue, a very small tweak could be decrypt only once in `commit` method and avoid decrypt twice (because `message_from` where you decrypt is called twice)

@fbrusatti thanks for the review! That's a good catch that the decryption happens multiple times. In my crude performance tests decryption was so quick the overhead didn't matter so I don't think I'll worry about tweaking it now. Let me know if you disagree and I can take a look at it.

**Comment by henare** [comment]
> I reckon 60-90ms overhead per payment isn't great, but acceptable. Maybe something that could be improved down the road. If it's multiples of 60-90ms per payment (not sure how many requests there are and how many of them need encryption) then maybe it's worth looking into it sooner than later.

Thanks @jscheid, it sounds like we can look at improving this down the road.

There are two requests per normal payment but these requests are separated by the OTP code being sent to the donor and then entered so the compounded overhead doesn't impact the experience.

**Review by fbrusatti — APPROVED** [review]

**Pull request merged**